swarm/network/kademlia.go

```diff
@@ -302,12 +304,24 @@ func (k *Kademlia) On(p OverlayConn) (uint8, bool) {
 	depth := uint8(k.saturation(k.MinBinSize))
 	var changed bool
 	if depth != k.depth {
+		// Non blocking send to channel.
+		// If there is no one to receive
+		// this method should not block.
+		select {
+		case k.depthC <- depth:
```

Comment:
the problem is 'no one to receive' can happen when the loop is processing the subscriptions. 
if we do this correctly later, we may not want to drop any change.
check if the channel is nil and create it only when a param is given. 
call the subscription requests async and make the channel buffered no?

Reply:
Thank you. This functionality is changed.